src/stdlib/replace_with.rs

```diff
@@ -0,0 +1,214 @@
+        replaced.push_str(&replacement);
+        last_match = m.end();
+        if i >= limit {
+            break;
+        }
+    }
+    // add the final component
+    replaced.push_str(&haystack[last_match..]);
+    Ok(replaced.into())
+}
+
+fn captures_to_value(captures: &regex::Captures) -> Value {
+    // return an array of the capture groups
+    captures
+        .iter()
+        .map(|m| match m {
+            Some(m) => m.as_str().into(),
+            None => Value::Null,
+        })
+        .collect()
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct ReplaceWith;
+
+impl Function for ReplaceWith {
+    fn identifier(&self) -> &'static str {
+        "replace_with"
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[
+            Parameter {
+                keyword: "value",
+                kind: kind::BYTES,
+                required: true,
+            },
+            Parameter {
+                keyword: "pattern",
+                kind: kind::REGEX,
+                required: true,
+            },
+            Parameter {
+                keyword: "count",
+                kind: kind::INTEGER,
+                required: false,
+            },
+        ]
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[
+            Example {
+                title: "double replacement",
+                source: r#"replace_with("foobar", r'o|a') -> |m| { m[0] + m[0] }"#,
+                result: Ok("foooobaar"),
+            },
+            Example {
+                title: "replace count",
+                source: r#"replace_with("foobar", r'o|a', count: 1) -> |m| { m[0] + m[0] }"#,
+                result: Ok("fooobar"),
+            },
+            Example {
+                title: "replace with capture group",
+                source: r#"replace_with("foo123bar", r'foo(\d+)bar') -> |m| { x = m[1]; "x={{x}}" }"#,
+                result: Ok(r#"x=123"#),
+            },
+            Example {
+                title: "process capture group",
+                source: r#"replace_with(s'Got message: {"msg": "b"}', r'message: (\{.*\})') -> |m| { to_string!(parse_json!(m[1]).msg) }"#,
+                result: Ok("Got b"),
+            },
+            Example {
+                title: "Optional capture group",
+                source: r#"replace_with("foobar", r'bar( of gold)?') -> |m| { if m[1] == null { "baz" } else { "rich" } }"#,
+                result: Ok("foobaz"),
+            },
+        ]
+    }
+
+    fn compile(
+        &self,
+        _state: &state::TypeState,
+        _ctx: &mut FunctionCompileContext,
+        arguments: ArgumentList,
+    ) -> Compiled {
+        let value = arguments.required("value");
+        let pattern = arguments.required("pattern");
+        let count = arguments.optional("count").unwrap_or(expr!(-1));
+
+        let closure = arguments.required_closure()?;
+
+        // TODO: Should I check that the closure is not fallible?
```

Comment:
Let's add some VRL tests (if unit tests for closures are a challenge). See this for example: https://github.com/vectordotdev/vrl/blob/main/lib/tests/tests/expressions/function_call/closure_scope_inheritance.vrl. You can just add ".vrl" files under `lib/tests/tests/*` and they will be discovered. Note that examples will also run as tests.

Reply:
I've added some.

Reply:
Regarding the TODO, I think we can remove it.

Reply:
removed